Execute remote actions on another extension

dbwiddis opened the pull request:
Companion PR: [#6734](https://github.com/opensearch-project/OpenSearch/pull/6734) must be merged first

### Description

Allows one extension to execute an action on another extension.

See the Hello World RestRemoteHelloAction for sample implementation.

Other work done to support:
 - Put a setter on Extension interface to get a copy of extensions runner since injection is flaky.
 - Created SDKTransportService class to avoid having to pass around so many params.   Lots of things can be moved from ExtensionsRunner there later, just did the action module ones in this PR and left some TODOs where other opportunities exist

### Issues Resolved

Fixes #525

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dbwiddis commented:
Gradle check is failing based on companion PR classes.  No idea why DCO check is failing, git log shows they're all signed.

dbwiddis commented on src/main/java/org/opensearch/sdk/SDKTransportService.java:
```diff
@@ -0,0 +1,126 @@
+     * @param request The request to send
+     * @return A buffer serializing the response from the remote action if successful, otherwise null
+     */
+    @Nullable
```
Great question, @vibrantvarun .  I'm working up a sequence diagram that should explain it, but for now here's a text-based version:
 - Anywhere in code that you want to call a remote action you will create a `ProxyActionRequest`.  In this PR you can start in the `RestRemoteHelloAction` but it doesn't need to be a REST call
 - The `ProxyActionRequest` contains 3 bits of information. If you have compile-time access to the classes, constructing it is simplified:
   - The remote action class name
   - The remote request class name
   - The remote request parameters as bytes to be used with the request's StreamInput constructor
 - `client.execute()` is called with the ProxyAction instance and the request generated in the previous step.  This causes the code in `ProxyTransportAction` to be invoked on the request.
 - `ProxyTransportAction.doExecute()` calls `sdkTransportService.sendProxyActionRequest(request)` (this method).
 - Sending the request (this method) copies info into a `TransportActionRequestFromExtension` which OpenSearch expects.  It will know from the action class name which extension has registered it and forwards it there.  It also glues together the request class name and request class bytes into a single byte array that is forwarded to the remote extension (in our hello world case it comes right back!)
 - On the remote extension, the `ExtensionActionRequestHandler` handles the inbound request.  See the `handleExtensionActionRequest()` method which has several comments showing how it finds the action to execute, reconstructs the appropriate request object from the bytes, runs it, and turns its response into bytes that get relayed back to OpenSearch and back to the originating extension.
 - Throughout this process there is exception handling
 - Back at the original sending point, if the bytes are received, they need to be used to construct the corresponding `Response` object for the original request and action.  This is done in the `RestRemoteHelloAction`.

dbwiddis replied:
@vibrantvarun see https://github.com/opensearch-project/opensearch-sdk-java/pull/588/files#diff-7e38c8f69d56fdabc2d8abee876b213b9a230a1b42a24239bf42c019a19e7aba

joshpalis commented on src/main/java/org/opensearch/sdk/SDKTransportService.java:
```diff
@@ -0,0 +1,126 @@
+import org.opensearch.transport.TransportService;
+
+/**
+ * Wrapper class for {@link TranspoortService} and associated methods.
```
Small nit, change to `{@link TansportService}`

owaiskazi19 commented on src/main/java/org/opensearch/sdk/handlers/ExtensionActionResponseHandler.java:
```diff
@@ -0,0 +1,82 @@
+    public void handleResponse(ExtensionActionResponse response) {
+        logger.info("Received response bytes: " + response.getResponseBytes().length + " bytes");
+
+        logger.debug("Received response bytes: " + response.getResponseBytesAsString());
```
Any specific reason to have info/debug together?

dbwiddis replied:
The bottom one was a debug string while things weren't working.  It's gone already.

owaiskazi19 commented on src/main/java/org/opensearch/sdk/handlers/ExtensionActionResponseHandler.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public void handleException(TransportException exp) {
+        logger.info("ExtensionActionResponseRequest failed", exp);
```
```suggestion
        logger.error("ExtensionActionResponseRequest failed", exp);
```

owaiskazi19 reviewed (commented):
Great work on Dynamic Actions @dbwiddis 🙌! Didn't get a chance to review the tests

saratvemulapalli commented on src/main/java/org/opensearch/sdk/handlers/ExtensionActionRequestHandler.java:
```diff
@@ -0,0 +1,140 @@
+import org.opensearch.sdk.action.SDKActionModule;
+
+/**
+ * This class handles a request from OpenSearch from another extension's {@link SDKTransportService#sendProxyActionRequest()} call.
```
Technically it doesnt really have to come from another extension, its just a transport request to the extension.

saratvemulapalli commented on src/main/java/org/opensearch/sdk/handlers/ExtensionActionRequestHandler.java:
```diff
@@ -0,0 +1,140 @@
+        final RemoteExtensionActionResponse response = new RemoteExtensionActionResponse(false, new byte[0]);
+
+        // Find matching ActionType instance
+        Optional<? extends ActionType<? extends ActionResponse>> optionalAction = sdkActionModule.getActions()
+            .values()
+            .stream()
+            .map(h -> h.getAction())
+            .filter(a -> a.getClass().getName().equals(request.getAction()))
+            .findAny();
+        if (optionalAction.isEmpty()) {
+            response.setResponseBytesAsString("No action [" + request.getAction() + "] is registered.");
+            return response;
+        }
```
I would like this to be a function of `SDKActionModule`. Sounds more like a utility.

dbwiddis replied:
See [this comment](https://github.com/opensearch-project/opensearch-sdk-java/pull/588#discussion_r1146694726). This has been replaced by a function that already existed on `SDKClient`.

dbwiddis commented on src/main/java/org/opensearch/sdk/ExtensionsRunner.java:
```diff
@@ -132,13 +134,15 @@ public class ExtensionsRunner {
     private ExtensionsIndicesModuleNameRequestHandler extensionsIndicesModuleNameRequestHandler =
         new ExtensionsIndicesModuleNameRequestHandler();
     private ExtensionsRestRequestHandler extensionsRestRequestHandler = new ExtensionsRestRequestHandler(extensionRestPathRegistry);
+    private ExtensionActionRequestHandler extensionsActionRequestHandler;
```
Sure, why not all of the ones around it at the same time?

dbwiddis commented on src/main/java/org/opensearch/sdk/ExtensionsRunner.java:
```diff
@@ -660,6 +684,8 @@ public static void run(Extension extension) throws IOException {
         // initialize the transport service
         NettyTransport nettyTransport = new NettyTransport(runner);
         runner.extensionTransportService = nettyTransport.initializeExtensionTransportService(runner.getSettings(), runner.getThreadPool());
+        // TODO: merge above line with below line when refactoring out extensionTransportService
```
Yup, https://github.com/opensearch-project/opensearch-sdk-java/issues/585

dbwiddis commented on src/main/java/org/opensearch/sdk/handlers/ExtensionActionRequestHandler.java:
```diff
@@ -0,0 +1,140 @@
+     * @return The response from the TransportAction
+     */
+    public ExtensionActionResponse handleExtensionActionRequest(ExtensionActionRequest request) {
+        logger.debug("Received request to execute action [" + request.getAction() + "]");
```
No, we presently are not forwarding any originating extension information. I'm not even sure we are able to do so unless the extension voluntarily provides it (like the unique Id in registering these actions to begin with) as all we really have is an incoming transport request.

dbwiddis commented on src/main/java/org/opensearch/sdk/SDKTransportService.java:
```diff
@@ -0,0 +1,126 @@
+
+    public void setUniqueId(String uniqueId) {
+        this.uniqueId = uniqueId;
+    }
```
This is part of the upcoming refactoring #585 and even mentioned there.  I had started doing it and decided not to clutter this PR with even more stuff once I saw how many files were using it.

dbwiddis commented on src/main/java/org/opensearch/sdk/action/ProxyAction.java:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * The name to look up this action with
+     */
+    public static final String NAME = "internal/proxyaction";
```
I already renamed this class and it's currently: `NAME = "internal/extension-proxyaction";`

dbwiddis replied:
Just changed it to `NAME = "internal:remote-extension-action";`

dbwiddis commented on src/main/java/org/opensearch/sdk/action/ProxyActionRequest.java:
```diff
@@ -0,0 +1,132 @@
+            bytes = BytesReference.toBytes(out.bytes());
+        } catch (IOException e) {
+            // This Should Never Happen (TM)
+            // Won't get an IOException locally
```
I'll throw an `IllegalStateException` per [this](https://stackoverflow.com/a/6271952/1161484)

dbwiddis commented on src/main/java/org/opensearch/sdk/action/ProxyTransportAction.java:
```diff
@@ -0,0 +1,57 @@
+        } else {
+            listener.onResponse(new ExtensionActionResponse(true, responseBytes));
+        }
+    }
```
Not sure what we're catching?  No exceptions are thrown from the code here.

dbwiddis commented on src/main/java/org/opensearch/sdk/action/SDKActionModule.java:
```diff
@@ -62,24 +53,33 @@ public ActionFilters getActionFilters() {
+         */
+        class ActionRegistry extends NamedRegistry<ActionHandler<?, ?>> {
+            ActionRegistry() {
+                super("sdkaction");
```
It really doesn't matter what we call it as it's a temporary registry with local scope, just to use easier register methods.  I'll stick it back to action.

dbwiddis commented on src/main/java/org/opensearch/sdk/action/SDKActionModule.java:
```diff
@@ -62,24 +53,33 @@ public ActionFilters getActionFilters() {
-        return Collections.emptyMap();
+        ActionRegistry actions = new ActionRegistry();
+
+        // Register SDK actions
```
These aren't necessarily remote actions.   They are just actions, in the SDK.

These are any permanent actions that we want to register in the SDK that any extension can use client.execute() with, as opposed to ones that extensions will register.   Similar to "core" actions vs. plugin actions in the corresponding module on openSearch.

Currently there's just one to execute remote actions, but there can conceivably be more as we implement more extension points.

dbwiddis commented on src/main/java/org/opensearch/sdk/handlers/ExtensionActionRequestHandler.java:
```diff
@@ -0,0 +1,140 @@
+
+        // Extract request class name from bytes and instantiate request
+        int nullPos = indexOf(request.getRequestBytes(), (byte) 0);
+        String requestClassName = new String(Arrays.copyOfRange(request.getRequestBytes(), 0, nullPos), StandardCharsets.UTF_8);
```
The +1 avoids index exceptions but leaves the resulting string with an extra null byte at the end.

So we could `trim()` or `stripTrailing()` on it, but those only handle [whitespace](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Character.html#isWhitespace(int)).

So, let's switch from the null delimiter byte to a whitespace byte.  The unicode `'\u001F', U+001F UNIT SEPARATOR.` looks like a good candidate.

dbwiddis commented on src/main/java/org/opensearch/sdk/SDKTransportService.java:
```diff
@@ -0,0 +1,126 @@
+        byte[] requestClassBytes = request.getRequestClass().getBytes(StandardCharsets.UTF_8);
+        byte[] proxyRequestBytes = ByteBuffer.allocate(requestClassBytes.length + 1 + request.getRequestBytes().length)
+            .put(requestClassBytes)
+            .put((byte) 0)
```
Per later comment, changed this to the unicode UNIT_DELIMITER character/byte.

dbwiddis commented:
@owaiskazi19 @saratvemulapalli addressed all your comments.

And as-of now it still works:
```
curl -X GET http://localhost:9200/_extensions/_hw/hello/test
Received greeting from remote extension: Hello, test%  
```

dbwiddis commented:
Gradle check failing due to companion PR.  Only thing left is to update the sequence diagram which I'm working on.

owaiskazi19 reviewed (approved):
LGTM!

joshpalis reviewed (approved):
Looks good to me, thanks @dbwiddis for this.
